### PR TITLE
Migrate security/reentrancyguard

### LINKED ISSUE #550: Migrate security libraries

### PR DESCRIPTION
**Pull request by andrew-fleming**
Fixes #550.

~~Finishing the tests is blocked by [call_contract_syscall](https://github.com/starkware-libs/cairo/blob/main/crates/cairo-lang-runner/src/casm_run/mod.rs#L410-L412) not yet being implemented in the testing environment. This is a requirement for recursive contract calls through the dispatchers.~~

One thing to note is the expected error message from reentrant calls include multiple `ENTRYPOINT_FAILED`s. 

#### PR Checklist

- [x] Tests
- [ ] Tried the feature on a public network
- [ ] Documentation

### CODE REVIEW

File: src/openzeppelin/mocks.cairo
Comment (martriay) [inline src/openzeppelin/mocks.cairo:1]: consider moving into `tests/`

File: src/openzeppelin/mocks/reentrancy_mock.cairo [hunk 1]
```diff
@@ -0,0 +1,85 @@
+#[abi]
+trait IReentrancyGuardAttacker {
+  fn call_sender();
+}
+
+#[abi]
+trait IReentrancyGuard {
```
Comment (martriay) [inline src/openzeppelin/mocks/reentrancy_mock.cairo:7]:
```suggestion
trait IReentrancyGuarded {
```

File: src/openzeppelin/mocks/reentrancy_mock.cairo [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        check_gas();
+        if n != 0 {
+            count();
+            count_local_recursive(n - 1);
+        }
+        ReentrancyGuard::end();
+    }
+
+    #[external]
+    fn count_this_recursive(n: felt252) {
```
Comment (martriay) [inline src/openzeppelin/mocks/reentrancy_mock.cairo:61]:
```suggestion
    fn count_this_recursive(n: felt252) {
```
should we rename this to something a bit more telling? like `count_external_recursive` or something

File: src/openzeppelin/security/reentrancyguard.cairo [hunk 3]
```diff
@@ -0,0 +1,17 @@
+#[contract]
+mod ReentrancyGuard {
+    use starknet::get_caller_address;
+
+    struct Storage {
+        _entered: bool,
```
Comment (martriay) [inline src/openzeppelin/security/reentrancyguard.cairo:6]:
```suggestion
        entered: bool,
```
same question/suggestion as in other PRs

File: src/openzeppelin/mocks/reentrancy_mock.cairo [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    struct Storage {
+        _counter: felt252
+    }
+
+    #[constructor]
+    fn constructor(initial_number: felt252) {
+        _counter::write(initial_number);
+    }
+
```
Comment (martriay) [inline src/openzeppelin/mocks/reentrancy_mock.cairo:36]: do we need a constructor and a parameter? can't we just rely on default behavior of `felt252`?

**Review by martriay — COMMENTED** [review]
looks good!

File: src/openzeppelin/mocks/reentrancy_mock.cairo (hunk 2 above)
Comment (andrew-fleming) [reply src/openzeppelin/mocks/reentrancy_mock.cairo:61]: Yeah, that's a much better name

**Review by martriay — COMMENTED** [review]
looking good 👍

File: src/openzeppelin/tests/test_reentrancyguard.cairo [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+//
+// Mock implementation tests
+//
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(
+    expected: (
+        'ReentrancyGuard: reentrant call',
+        'ENTRYPOINT_FAILED',
```
Comment (ericnordelo) [inline src/openzeppelin/tests/test_reentrancyguard.cairo:71]: Why do we get three ENTRYPOINT_FAILED? Is one for each call in the stack of calls?

**Review by ericnordelo — COMMENTED** [review]
Looking great! I left a couple of small suggestions.

File: src/openzeppelin/tests/mocks/reentrancy_mock.cairo [hunk 6]
```diff
@@ -0,0 +1,100 @@
+use starknet::ContractAddress;
+
+#[abi]
+trait IAttacker {
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/tests/mocks/reentrancy_mock.cairo:4]: Good call!

File: src/openzeppelin/tests/test_reentrancyguard.cairo (hunk 5 above)
Comment (andrew-fleming) [reply src/openzeppelin/tests/test_reentrancyguard.cairo:71]: > Is one for each call in the stack of calls?

That's my understanding of it, yes

File: src/openzeppelin/tests/test_reentrancyguard.cairo [hunk 7]
```diff
@@ -1,43 +1,107 @@
...
 
+use array::ArrayTrait;
+use core::result::ResultTrait;
+use option::OptionTrait;
+use starknet::class_hash::Felt252TryIntoClassHash;
 use starknet::ContractAddress;
+use traits::TryInto;
+
+fn deploy_mock() -> IReentrancyMockDispatcher {
+    let calldata = ArrayTrait::<felt252>::new();
```
Comment (martriay) [inline src/openzeppelin/tests/test_reentrancyguard.cairo:15]:
```suggestion
    let calldata = ArrayTrait::new();
```
the compiler is smart enough

File: src/openzeppelin/tests/test_reentrancyguard.cairo [hunk 8]
```diff
@@ -1,43 +1,107 @@
...
+    let (address, _) = starknet::deploy_syscall(
+        ReentrancyMock::TEST_CLASS_HASH.try_into().unwrap(), 0, calldata.span(), false
+    )
```
Comment (martriay) [inline src/openzeppelin/tests/test_reentrancyguard.cairo:18]: thinking we could abstract deployments to make them easier to read?

```suggestion
    let address = deploy(ReentrancyMock, calldata);
```

```cairo
fn deploy(contract, calldata) -> ContractAddress {
    let (address, _) = starknet::deploy_syscall(
        contract::TEST_CLASS_HASH.try_into().unwrap(), 0, calldata.span(), false
    )
    address
}
```

or something similar

File: src/openzeppelin/tests/test_reentrancyguard.cairo [hunk 9]
```diff
@@ -1,43 +1,107 @@
...
+    let (address, _) = starknet::deploy_syscall(
+        Attacker::TEST_CLASS_HASH.try_into().unwrap(), 0, calldata.span(), false
+    )
+        .unwrap();
+
+    address
+}
+
+//
+// ReentrancyGuard direct call tests
```
Comment (martriay) [inline src/openzeppelin/tests/test_reentrancyguard.cairo:35]: nice addition

**Review by martriay — COMMENTED** [review]
great work

File: src/openzeppelin/tests/test_reentrancyguard.cairo (hunk 8 above)
Comment (andrew-fleming) [reply src/openzeppelin/tests/test_reentrancyguard.cairo:18]: Great idea!

**Review by martriay — APPROVED** [review]

**Pull request merged**